crates/bevy_ecs/src/query/filter.rs

```diff
@@ -445,20 +458,118 @@ macro_rules! impl_query_filter_tuple {
 
 all_tuples!(impl_query_filter_tuple, 0, 15, F, S);
 
-macro_rules! impl_flag_filter {
+/// Query transformer that inverses its inner filter.
+/// For example, `Not<With<T>>` is equivalent to `Without<T>`
+pub struct Not<T>(pub T);
+
+pub struct NotFetch<T: FilterFetch> {
+    fetch: T,
+    matches: bool,
+}
+
+impl<T: WorldQuery> WorldQuery for Not<T>
+where
+    T::Fetch: FilterFetch,
+{
+    type Fetch = Not<NotFetch<T::Fetch>>;
+    type State = Not<T::State>;
+}
+
+impl<'a, T: FilterFetch> Fetch<'a> for Not<NotFetch<T>> {
+    type Item = bool;
+    type State = Not<<T as Fetch<'a>>::State>;
+
+    unsafe fn init(
+        world: &World,
+        state: &Self::State,
+        system_counter: u32,
+        global_system_counter: u32,
+    ) -> Self {
+        Not(NotFetch {
+            fetch: T::init(world, &state.0, system_counter, global_system_counter),
+            matches: false,
+        })
+    }
+
+    fn is_dense(&self) -> bool {
+        self.0.fetch.is_dense()
+    }
+
+    unsafe fn set_table(&mut self, state: &Self::State, table: &Table) {
+        self.0.matches = state.0.matches_table(table);
+        if self.0.matches {
+            self.0.fetch.set_table(&state.0, table);
+        }
+    }
+
+    unsafe fn set_archetype(
+        &mut self,
+        state: &Self::State,
+        archetype: &Archetype,
+        tables: &Tables,
+    ) {
+        self.0.matches = state.0.matches_archetype(archetype);
+        if self.0.matches {
+            self.0.fetch.set_archetype(&state.0, archetype, tables);
+        }
+    }
+
+    unsafe fn table_fetch(&mut self, table_row: usize) -> bool {
+        dbg!((&self.0.matches, self.0.fetch.table_filter_fetch(table_row)));
+        !(self.0.matches && self.0.fetch.table_filter_fetch(table_row))
+    }
+
+    unsafe fn archetype_fetch(&mut self, archetype_index: usize) -> bool {
+        dbg!((
+            &self.0.matches,
+            self.0.fetch.archetype_filter_fetch(archetype_index)
+        ));
+        !(self.0.matches && self.0.fetch.archetype_filter_fetch(archetype_index))
+    }
+}
+
+unsafe impl<T: FetchState> FetchState for Not<T> {
+    fn init(world: &mut World) -> Self {
+        Not(T::init(world))
+    }
+
+    fn update_component_access(&self, access: &mut FilteredAccess<ComponentId>) {
+        self.0.update_component_access(access);
+    }
+
+    fn update_archetype_component_access(
+        &self,
+        archetype: &Archetype,
+        access: &mut Access<ArchetypeComponentId>,
+    ) {
+        self.0.update_archetype_component_access(archetype, access);
+    }
+
+    fn matches_archetype(&self, _archetype: &Archetype) -> bool {
+        true
```

Comment:
I tried, but couldn't make it work that way. Returning false here short-circuits, but `Not` can match both things that do or don't match the archetype. For instance, `Not<Added<T>>` should match both entities without `T` and entities which `T` was not just added (I may be wrong there, but that's my current understanding).

Reply:
> For instance, Not<Added<T>> should match both entities without T and entities which T was not just added

Right. That behavior is unintuitive at first glance, but you don't want to accidentally cause `Not<Added<T>>` to act like `(Not<Added<T>>, With<T>)`, since that combines filtering with the obvious "it hasn't had a `T` added behavior in a way that I don't think is simply reversed.

Reply:
Hmm yeah that makes sense. Maybe we should leave `Not<T>` out of this pr (and maybe 0.5). It feels like a bit of a performance foot-gun right now. Users that absolutely need `Not<Changed<T>>` can always manually filter on `ChangeTrackers` (which isn't ideal ergonomically but does ensure that `Not<T>` cant be misused).

Reply:
> Maybe we should leave Not<T> out of this pr (and maybe 0.5)

I agree here: the use cases aren't very clear, and it's mostly just a foot gun for both performance and clarity. The impetus for adding it was to allow easy reproduction of `Mutated` behavior, but no one was ever using that anyways.

Until we have a clear (even if niche) use case, I think it should just be left out. Once there's a use case, include it for logical completeness of the standard Or / And / Not behavior.

Reply:
Since the main use case for Not is optimizations, I agree. I hadn't realized the performance implications. Removed in 34ec8eaa31e64756c4bb7514ac0e6b9b1b8c85b7.